crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,498 @@
+            _,
+        ) => {
+            func.as_name_expr()
+                .is_some_and(|name| name.id == ast::name::Name::new("len"))
+                && arguments.len() == 1
+                && arguments.find_positional(0).is_some_and(|arg| {
+                    let compr_affix = ast::comparable::ComparableExpr::from(affix);
+                    let compr_arg = ast::comparable::ComparableExpr::from(arg);
+                    compr_affix == compr_arg
+                })
+        }
+        (
+            AffixKind::EndsWith,
+            ast::Expr::UnaryOp(ast::ExprUnaryOp {
+                op: ast::UnaryOp::USub,
+                operand,
+                range: _,
+            }),
+            ast::Expr::StringLiteral(ast::ExprStringLiteral {
+                range: _,
+                value: string_val,
+            }),
+        ) => operand.as_number_literal_expr().is_some_and(
+            |ast::ExprNumberLiteral { value, .. }| {
+                value.as_int().is_some_and(|x| {
+                    let length = u32::try_from(string_val.len())
+                        // Only support suffix removal for size at most `u32::MAX`.
+                        .unwrap_or(u32::MAX);
+                    x == &ast::Int::from(length)
+                })
+            },
+        ),
+        (
+            AffixKind::EndsWith,
+            ast::Expr::UnaryOp(ast::ExprUnaryOp {
+                op: ast::UnaryOp::USub,
+                operand,
+                range: _,
+            }),
+            _,
+        ) => operand.as_call_expr().is_some_and(
+            |ast::ExprCall {
+                 range: _,
+                 func,
+                 arguments,
+             }| {
+                func.as_name_expr()
+                    .is_some_and(|name| name.id == ast::name::Name::new("len"))
+                    && arguments.len() == 1
+                    && arguments.find_positional(0).is_some_and(|arg| {
+                        let compr_affix = ast::comparable::ComparableExpr::from(affix);
+                        let compr_arg = ast::comparable::ComparableExpr::from(arg);
+                        compr_affix == compr_arg
+                    })
+            },
+        ),
+        _ => false,
+    }
+}
+
+/// Generates the source code string
+/// ```python
+/// text = text.removeprefix(prefix)
+/// ```
+/// or
+/// ```python
+/// text = text.removesuffix(prefix)
+/// ```
+/// as appropriate.
+fn generate_assignment_with_removeaffix(
+    text: &str,
+    affix_query: &AffixQuery,
+    generator: Generator,
+) -> String {
+    let remove_affix_expr = make_removeaffix_expr(text, affix_query);
+    generator.stmt(&ast::Stmt::Assign(ast::StmtAssign {
+        range: TextRange::default(),
+        targets: vec![ast::Expr::Name(ast::ExprName {
+            range: TextRange::default(),
+            id: ast::name::Name::from(text),
+            ctx: ast::ExprContext::Store,
+        })],
+        value: Box::new(remove_affix_expr),
+    }))
+}
```

Comment:
I think you might be able to do this more simply (and probably more cheaply) using some of the techniques I showed you in https://github.com/astral-sh/ruff/pull/13275 (string formatting, using `checker.locator().slice(<expr_node>)` to get the node as it exists in the source)

Reply:
Great idea! Implemented